File: Samples/CS/CodeSamples/DeliveryReceiptSample.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Inetlab.SMPP;
using Inetlab.SMPP.Common;
using Inetlab.SMPP.PDU;

namespace CodeSamples
{
    public class DeliveryReceiptSample
    {
        readonly SmppClient _client = new SmppClient();

        private readonly SmppConfig _config;
        private readonly IClientMessageStore _clientMessageStore;

        public DeliveryReceiptSample(SmppConfig config, IClientMessageStore clientMessageStore)
        {
            _config = config;
            _clientMessageStore = clientMessageStore;
            _client.evDeliverSm += ClientOnEvDeliverSm;
        }

        //<SendMessage>
        public async Task SendMessage(TextMessage message)
        {
            IList<SubmitSm> list = SMS.ForSubmit()
                .From(_config.ShortCode)
                .To(message.PhoneNumber)
                .Text(message.Text)
                .DeliveryReceipt()
                .Create(_client);

            foreach (SubmitSm sm in list)
            {
                sm.Header.Sequence = _client.SequenceGenerator.NextSequenceNumber();
                _clientMessageStore.SaveSequence(message.Id, sm.Header.Sequence);
            }

          var responses = await _client.Submit(list);

            foreach (SubmitSmResp resp in responses)
            {
                _clientMessageStore.SaveMessageId(message.Id, resp.MessageId);
            }
        }
        //</SendMessage>


        //<EvDeliverSm>
        private void ClientOnEvDeliverSm(object sender, DeliverSm data)
        {
            if (data.MessageType == MessageTypes.SMSCDeliveryReceipt)
            {
                _clientMessageStore.UpdateMessageStatus(data.Receipt.MessageId, data.Receipt.State);
            }
        }
        //</EvDeliverSm>
        // 
        public interface IClientMessageStore
        {
            TextMessage GetNextMessage();

            void SaveSequence(string id, uint sequence);
            void SaveMessageId(string id, string smppMessageId);
            void UpdateMessageStatus(string smppMessageId, MessageState state);
        }

        public class TextMessage
        {
            public string Id { get; set; }
            public string PhoneNumber { get; set; }
            public string Text { get; set; }

            public string ServiceAddress { get; set; }
        }
    }

  
}

File: Samples/VB.NET/CodeSamples/DeliveryReceiptSample.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Threading.Tasks
Imports Inetlab.SMPP
Imports Inetlab.SMPP.Common
Imports Inetlab.SMPP.PDU

Namespace CodeSamples
	Public Class DeliveryReceiptSample
		Private ReadOnly _client As New SmppClient()

		Private ReadOnly _config As SmppConfig
		Private ReadOnly _clientMessageStore As IClientMessageStore

		Public Sub New(ByVal config As SmppConfig, ByVal clientMessageStore As IClientMessageStore)
			_config = config
			_clientMessageStore = clientMessageStore
			AddHandler _client.evDeliverSm, AddressOf ClientOnEvDeliverSm
		End Sub

		'<SendMessage>
		Public Async Function SendMessage(ByVal message As TextMessage) As Task
			Dim list As IList(Of SubmitSm) = SMS.ForSubmit().From(_config.ShortCode).To(message.PhoneNumber).Text(message.Text).DeliveryReceipt().Create(_client)

			For Each sm As SubmitSm In list
				sm.Header.Sequence = _client.SequenceGenerator.NextSequenceNumber()
				_clientMessageStore.SaveSequence(message.Id, sm.Header.Sequence)
			Next sm

		  Dim responses = Await _client.Submit(list)

			For Each resp As SubmitSmResp In responses
				_clientMessageStore.SaveMessageId(message.Id, resp.MessageId)
			Next resp
		End Function
		'</SendMessage>


		'<EvDeliverSm>
		Private Sub ClientOnEvDeliverSm(ByVal sender As Object, ByVal data As DeliverSm)
			If data.MessageType = MessageTypes.SMSCDeliveryReceipt Then
				_clientMessageStore.UpdateMessageStatus(data.Receipt.MessageId, data.Receipt.State)
			End If
		End Sub
		'</EvDeliverSm>
		' 
		Public Interface IClientMessageStore
			Function GetNextMessage() As TextMessage

			Sub SaveSequence(ByVal id As String, ByVal sequence As UInteger)
			Sub SaveMessageId(ByVal id As String, ByVal smppMessageId As String)
			Sub UpdateMessageStatus(ByVal smppMessageId As String, ByVal state As MessageState)
		End Interface

		Public Class TextMessage
			Public Property Id() As String
			Public Property PhoneNumber() As String
			Public Property Text() As String

			Public Property ServiceAddress() As String
		End Class
	End Class


End Namespace
